Spurious update to opensearch_cluster_settings yields 401

https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/92841

```
aws_lambda_function.service_manifest: Modifications complete after 5s [id=azul-service-dev-manifest]
╷
│ Error: elastic: Error 401 (Unauthorized)
│ 
│   with opensearch_cluster_settings.index,
│   on elasticsearch.tf.json line 153, in resource[5].opensearch_cluster_settings[0].index:
│  153:                     }
│ 
╵
make[1]: *** [Makefile:64: auto_apply] Error 1
make[1]: Leaving directory '/builds/ucsc/azul/terraform'
make: *** [Makefile:106: auto_terraform] Error 2
Cleaning up project directory and file based variables 00:01
ERROR: Job failed: exit code 1
```
Re-running the job did not resolve the problem,
 https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/92856

--- a/scripts/manage_cluster_settings.py
+++ b/scripts/manage_cluster_settings.py
@@ -1,0 +1,25 @@
+from azul import (
+    require,
+)
+from azul.es import (
+    ESClientFactory,
+)
+from azul.logging import (
+    configure_script_logging,
+)
+
+configure_script_logging()
+
+
+def main():
+    es = ESClientFactory.get()
+    response = es.cluster.put_settings(body={
+        'persistent': {
+            'action.auto_create_index': False
+        }
+    })
+    require(response['acknowledged'], 'Failed to update cluster settings', response)
+
+
+if __name__ == '__main__':
+    main()

--- a/scripts/rename_resources.py
+++ b/scripts/rename_resources.py
@@ -19,6 +19,9 @@ from azul.terraform import (
 log = logging.getLogger(__name__)
 
 renamed: dict[str, Optional[str]] = {
+    # FIXME: Remove the following entry
+    #        https://github.com/DataBiosphere/azul/pull/6285
+    'opensearch_cluster_settings.index': None
 }
 
 

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -784,11 +784,13 @@ class Chalice:
         # and less intrusive fix.
         #
         deployment = resources['aws_api_gateway_deployment'][app_name]
+        stage_name = deployment.pop('stage_name')
+        require(stage_name == config.deployment_stage,
+                'The TF config from Chalice does not match the selected deployment',
+                stage_name, config.deployment_stage)
         del deployment['lifecycle']['create_before_destroy']
         assert not deployment['lifecycle'], deployment
         del deployment['lifecycle']
-        stage_name = deployment.pop('stage_name')
-        assert stage_name == config.deployment_stage, stage_name
         deployment['triggers'] = {'redeployment': deployment.pop('stage_description')}
 
         return {

--- a/terraform/Makefile
+++ b/terraform/Makefile
@@ -34,18 +34,18 @@ validate: config
 	terraform validate
 
 .PHONY: rename_resources
-rename_resources: config
+rename_resources: validate
 	python $(project_root)/scripts/rename_resources.py
 
 .PHONY: import_resources
 import_resources: rename_resources
 
 .PHONY: plan
-plan: validate import_resources
+plan: import_resources
 	terraform plan
 
 .PHONY: apply
-apply: validate import_resources
+apply: import_resources
 ifeq ($(AZUL_PRIVATE_API),1)
 	# For private API we need the VPC endpoints to be created first so that the
 	# aws_lb_target_group_attachment can iterate over the network_interface_ids.
@@ -54,7 +54,7 @@ endif
 	terraform apply
 
 .PHONY: auto_apply
-auto_apply: validate import_resources
+auto_apply: import_resources
 ifeq ($(AZUL_PRIVATE_API),1)
 	# See `apply` above
 	terraform apply -auto-approve -target aws_vpc_endpoint.indexer -target aws_vpc_endpoint.service
@@ -64,11 +64,11 @@ endif
 	terraform apply -auto-approve plan.bin
 
 .PHONY: destroy
-destroy: validate import_resources
+destroy: import_resources
 	terraform destroy
 
 .PHONY: auto_destroy
-auto_destroy: validate import_resources
+auto_destroy: import_resources
 	@echo '!!! All resources will be deleted in 10s, hit Ctrl-C to cancel !!!'; sleep 10
 	terraform destroy -auto-approve
 

--- a/terraform/Makefile
+++ b/terraform/Makefile
@@ -4,10 +4,10 @@ all: apply
 include ../common.mk
 
 .PHONY: clean
-clean: check_env git_clean
+clean: check_env check_branch git_clean
 
 .PHONY: state
-state: check_terraform check_branch check_aws
+state: check_terraform check_aws
 
 .PHONY: initable
 initable: clean state providers.tf.json backend.tf.json

--- a/terraform/Makefile
+++ b/terraform/Makefile
@@ -14,8 +14,7 @@ initable: clean state providers.tf.json backend.tf.json
 
 .PHONY: init
 init: initable
-	terraform init -reconfigure
-	$(MAKE) check_providers_clean
+	terraform init -reconfigure -lockfile=readonly
 
 .PHONY: check_schema
 check_schema: init
@@ -79,8 +78,3 @@ provider_update: initable
 	                         -platform=linux_arm64 \
 	                         -platform=darwin_amd64 \
 	                         -platform=darwin_arm64
-
-.PHONY: check_providers_clean
-check_providers_clean:
-	git diff --exit-code --name-only .terraform.lock.hcl \
-	&& git diff --cached --exit-code --name-only .terraform.lock.hcl

--- a/terraform/elasticsearch.tf.json.template.py
+++ b/terraform/elasticsearch.tf.json.template.py
@@ -191,6 +191,6 @@ emit_tf(None if config.share_es_domain else {
                     ],
                 }
             }
-        } if domain else {}
+        }
     ]
 })

--- a/terraform/elasticsearch.tf.json.template.py
+++ b/terraform/elasticsearch.tf.json.template.py
@@ -24,34 +24,34 @@ emit_tf(None if config.share_es_domain else {
             }
         }
     ],
-    "resource": [
+    'resource': [
         *({
-            "aws_cloudwatch_log_group": {
-                f"{log}_log": {
-                    "name": f"/aws/aes/domains/{domain}/{log}-logs",
-                    "retention_in_days": config.audit_log_retention_days
+            'aws_cloudwatch_log_group': {
+                f'{log}_log': {
+                    'name': f'/aws/aes/domains/{domain}/{log}-logs',
+                    'retention_in_days': config.audit_log_retention_days
                 }
             }
         } for log in logs.keys()),
         {
-            "aws_cloudwatch_log_resource_policy": {
-                "index": {
-                    "policy_name": domain,
-                    "policy_document": json.dumps(
+            'aws_cloudwatch_log_resource_policy': {
+                'index': {
+                    'policy_name': domain,
+                    'policy_document': json.dumps(
                         {
-                            "Version": "2012-10-17",
-                            "Statement": [
+                            'Version': '2012-10-17',
+                            'Statement': [
                                 {
-                                    "Effect": "Allow",
-                                    "Principal": {
-                                        "Service": "es.amazonaws.com"
+                                    'Effect': 'Allow',
+                                    'Principal': {
+                                        'Service': 'es.amazonaws.com'
                                     },
-                                    "Action": [
-                                        "logs:PutLogEvents",
-                                        "logs:CreateLogStream"
+                                    'Action': [
+                                        'logs:PutLogEvents',
+                                        'logs:CreateLogStream'
                                     ],
-                                    "Resource": [
-                                        "${aws_cloudwatch_log_group." + log + "_log.arn}:*" for log in logs.keys()
+                                    'Resource': [
+                                        '${aws_cloudwatch_log_group.' + log + '_log.arn}:*' for log in logs.keys()
                                     ]
                                 }
                             ]
@@ -61,62 +61,62 @@ emit_tf(None if config.share_es_domain else {
             }
         },
         {
-            "aws_elasticsearch_domain": {
-                "index": {
-                    "access_policies": json.dumps({
-                        "Version": "2012-10-17",
-                        "Statement": [
+            'aws_elasticsearch_domain': {
+                'index': {
+                    'access_policies': json.dumps({
+                        'Version': '2012-10-17',
+                        'Statement': [
                             {
-                                "Effect": "Allow",
-                                "Principal": {
-                                    "AWS": "arn:aws:iam::${local.account_id}:root"
+                                'Effect': 'Allow',
+                                'Principal': {
+                                    'AWS': 'arn:aws:iam::${local.account_id}:root'
                                 },
-                                "Action": "es:*",
-                                "Resource": "arn:aws:es:${local.region}:${local.account_id}:domain/" + domain + "/*"
+                                'Action': 'es:*',
+                                'Resource': 'arn:aws:es:${local.region}:${local.account_id}:domain/' + domain + '/*'
                             }
                         ]
                     }),
-                    "advanced_options": {
-                        "rest.action.multi.allow_explicit_index": "true",
-                        "override_main_response_version": "false"
+                    'advanced_options': {
+                        'rest.action.multi.allow_explicit_index': 'true',
+                        'override_main_response_version': 'false'
                     },
-                    "cluster_config": {
-                        "instance_count": config.es_instance_count,
-                        "instance_type": config.es_instance_type,
+                    'cluster_config': {
+                        'instance_count': config.es_instance_count,
+                        'instance_type': config.es_instance_type,
                         # Needed for using multiple subnets (1 per zone) in the VPC
-                        "zone_awareness_enabled": True
+                        'zone_awareness_enabled': True
                     },
-                    "domain_name": domain,
-                    "ebs_options": {
-                        "ebs_enabled": "true",
-                        "volume_size": config.es_volume_size,
-                        "volume_type": "gp2"
+                    'domain_name': domain,
+                    'ebs_options': {
+                        'ebs_enabled': 'true',
+                        'volume_size': config.es_volume_size,
+                        'volume_type': 'gp2'
                     } if config.es_volume_size else {
-                        "ebs_enabled": "false",
+                        'ebs_enabled': 'false',
                     },
-                    "vpc_options": {
-                        "subnet_ids": [
-                            f"${{data.aws_subnet.gitlab_private_{zone}.id}}"
+                    'vpc_options': {
+                        'subnet_ids': [
+                            f'${{data.aws_subnet.gitlab_private_{zone}.id}}'
                             for zone in range(vpc.num_zones)
                         ],
-                        "security_group_ids": ["${aws_security_group.elasticsearch.id}"]
+                        'security_group_ids': ['${aws_security_group.elasticsearch.id}']
                     },
-                    "elasticsearch_version": "7.10",
-                    "log_publishing_options": [
+                    'elasticsearch_version': '7.10',
+                    'log_publishing_options': [
                         {
-                            "cloudwatch_log_group_arn": "${aws_cloudwatch_log_group." + log + "_log.arn}",
-                            "enabled": "true" if enabled else "false",
-                            "log_type": log_type
+                            'cloudwatch_log_group_arn': '${aws_cloudwatch_log_group.' + log + '_log.arn}',
+                            'enabled': 'true' if enabled else 'false',
+                            'log_type': log_type
                         } for log, (log_type, enabled) in logs.items()
                     ],
-                    "lifecycle": {
-                        "ignore_changes": [
+                    'lifecycle': {
+                        'ignore_changes': [
                             # Quoting AWS support:
                             #
                             # > Please note that with automated snapshots
-                            # > disabled, the "automatedSnapshotStartHour"
+                            # > disabled, the 'automatedSnapshotStartHour'
                             # > parameter of the domain configuration is set
-                            # > to "-1" from the service end (this can only
+                            # > to '-1' from the service end (this can only
                             # > be done from the service side). Please ensure
                             # > that this parameter is not overriden to a
                             # > different value from your end, else the
@@ -134,19 +134,19 @@ emit_tf(None if config.share_es_domain else {
                             # safer to me. The property is deprecated
                             # anyways.
                             #
-                            "snapshot_options"
+                            'snapshot_options'
                         ]
                     },
-                    "encrypt_at_rest": {
-                        "enabled": True
+                    'encrypt_at_rest': {
+                        'enabled': True
                     },
-                    "node_to_node_encryption": {
-                        "enabled": True
+                    'node_to_node_encryption': {
+                        'enabled': True
                     }
                 }
             },
-            "opensearch_cluster_settings": {
-                "index": {
+            'opensearch_cluster_settings': {
+                'index': {
                     # Disable the automatic creation of indexes when documents
                     # are indexed. We create indexes explicitly before any
                     # documents are indexed so a missing index would be
@@ -164,22 +164,22 @@ emit_tf(None if config.share_es_domain else {
                     #
                     # https://github.com/opensearch-project/terraform-provider-opensearch/issues/60#issuecomment-2041280397
                     #
-                    "action_auto_create_index": False,
-                    "lifecycle": {
-                        "ignore_changes": [
-                            "cluster_routing_allocation_disk_watermark_low",
-                            "cluster_routing_allocation_disk_watermark_high"
+                    'action_auto_create_index': False,
+                    'lifecycle': {
+                        'ignore_changes': [
+                            'cluster_routing_allocation_disk_watermark_low',
+                            'cluster_routing_allocation_disk_watermark_high'
                         ]
                     }
                 }
             },
-            "aws_security_group": {
-                "elasticsearch": {
-                    "name": config.qualified_resource_name("elasticsearch"),
-                    "vpc_id": "${data.aws_vpc.gitlab.id}",
-                    "ingress": [
-                        vpc.security_rule(cidr_blocks=["${data.aws_vpc.gitlab.cidr_block}"],
-                                          protocol="tcp",
+            'aws_security_group': {
+                'elasticsearch': {
+                    'name': config.qualified_resource_name('elasticsearch'),
+                    'vpc_id': '${data.aws_vpc.gitlab.id}',
+                    'ingress': [
+                        vpc.security_rule(cidr_blocks=['${data.aws_vpc.gitlab.cidr_block}'],
+                                          protocol='tcp',
                                           from_port=443,
                                           to_port=443)
                     ],

--- a/terraform/elasticsearch.tf.json.template.py
+++ b/terraform/elasticsearch.tf.json.template.py
@@ -17,13 +17,6 @@ logs = {
 domain = config.es_domain
 
 emit_tf(None if config.share_es_domain else {
-    'provider': [
-        {
-            'opensearch': {
-                'url': 'https://${aws_elasticsearch_domain.index.endpoint}'
-            }
-        }
-    ],
     'resource': [
         *(
             {
@@ -151,31 +144,26 @@ emit_tf(None if config.share_es_domain else {
                     }
                 }
             },
-            'opensearch_cluster_settings': {
-                'index': {
-                    # Disable the automatic creation of indexes when documents
-                    # are indexed. We create indexes explicitly before any
-                    # documents are indexed so a missing index would be
-                    # indicative of some sort of bug. We want to fail early in
-                    # that situation. Automatically created indices have a only
-                    # a default mapping, resulting in failure modes that are
-                    # harder to diagnose.
-                    #
-                    # The default of this setting is True. Due to a bug in the
-                    # TF provider, modifying this setting, or any cluster
-                    # setting for that matter, requires removing the state for
-                    # the resource prior to applying the modified
-                    # configuration. The ignore_changes property below is also
-                    # part of the workaround for the bug.
-                    #
-                    # https://github.com/opensearch-project/terraform-provider-opensearch/issues/60#issuecomment-2041280397
-                    #
-                    'action_auto_create_index': False,
+            'null_resource': {
+                'cluster_settings': {
+                    'depends_on': [
+                        'aws_elasticsearch_domain.index'
+                    ],
+                    'triggers': {
+                        'script_hash': '${filesha256("%s/scripts/manage_cluster_settings.py")}' % config.project_root
+                    },
                     'lifecycle': {
-                        'ignore_changes': [
-                            'cluster_routing_allocation_disk_watermark_low',
-                            'cluster_routing_allocation_disk_watermark_high'
+                        'replace_triggered_by': [
+                            'aws_elasticsearch_domain.index'
                         ]
+                    },
+                    'provisioner': {
+                        'local-exec': {
+                            'command': ' '.join([
+                                'python',
+                                f'{config.project_root}/scripts/manage_cluster_settings.py'
+                            ]),
+                        }
                     }
                 }
             },

--- a/terraform/elasticsearch.tf.json.template.py
+++ b/terraform/elasticsearch.tf.json.template.py
@@ -25,14 +25,17 @@ emit_tf(None if config.share_es_domain else {
         }
     ],
     'resource': [
-        *({
-            'aws_cloudwatch_log_group': {
-                f'{log}_log': {
-                    'name': f'/aws/aes/domains/{domain}/{log}-logs',
-                    'retention_in_days': config.audit_log_retention_days
+        *(
+            {
+                'aws_cloudwatch_log_group': {
+                    f'{log}_log': {
+                        'name': f'/aws/aes/domains/{domain}/{log}-logs',
+                        'retention_in_days': config.audit_log_retention_days
+                    }
                 }
             }
-        } for log in logs.keys()),
+            for log in logs.keys()
+        ),
         {
             'aws_cloudwatch_log_resource_policy': {
                 'index': {
@@ -91,7 +94,9 @@ emit_tf(None if config.share_es_domain else {
                         'ebs_enabled': 'true',
                         'volume_size': config.es_volume_size,
                         'volume_type': 'gp2'
-                    } if config.es_volume_size else {
+                    }
+                    if config.es_volume_size else
+                    {
                         'ebs_enabled': 'false',
                     },
                     'vpc_options': {
@@ -107,7 +112,8 @@ emit_tf(None if config.share_es_domain else {
                             'cloudwatch_log_group_arn': '${aws_cloudwatch_log_group.' + log + '_log.arn}',
                             'enabled': 'true' if enabled else 'false',
                             'log_type': log_type
-                        } for log, (log_type, enabled) in logs.items()
+                        }
+                        for log, (log_type, enabled) in logs.items()
                     ],
                     'lifecycle': {
                         'ignore_changes': [

--- a/terraform/providers.tf.json.template.py
+++ b/terraform/providers.tf.json.template.py
@@ -25,6 +25,8 @@ emit_tf(tag_resources=False, config={
                 'source': 'hashicorp/aws',
                 'version': '5.11.0'
             },
+            # FIXME: Remove the provider
+            #        https://github.com/DataBiosphere/azul/pull/6285
             'opensearch': {
                 'source': 'opensearch-project/opensearch',
                 'version': '2.2.1'

